stl/inc/xutility

```diff
@@ -6070,6 +6072,114 @@ private:
     _Udiff _Bmask; // 2^_Bits - 1
 };
 
+template <class _Diff, class _Urng>
+class _Rng_from_urng_v2 { // wrap a URNG as an RNG
+public:
+    using _Ty0 = make_unsigned_t<_Diff>;
+    using _Ty1 = _Invoke_result_t<_Urng&>;
+
+    using _Udiff                              = conditional_t<sizeof(_Ty1) < sizeof(_Ty0), _Ty0, _Ty1>;
+    static constexpr unsigned int _Udiff_bits = sizeof(_Udiff) * CHAR_BIT;
+    using _Uprod = conditional_t<_Udiff_bits <= 16, uint32_t, conditional_t<_Udiff_bits <= 32, uint64_t, _Unsigned128>>;
+
+    explicit _Rng_from_urng_v2(_Urng& _Func) : _Ref(_Func) {}
+
+    _Diff operator()(_Diff _Index) { // adapt _Urng closed range to [0, _Index)
+        // From Daniel Lemire, "Fast Random Integer Generation in an Interval", ACM Trans. Model. Comput. Simul. 29 (1),
+        // 2019.
+        _Udiff _Mask        = _Bmask;
+        unsigned int _Niter = 1;
+
+        if constexpr (_Bits < _Udiff_bits) {
+            while (_Mask < static_cast<_Udiff>(_Index - 1)) {
+                _Mask <<= _Bits;
+                _Mask |= _Bmask;
+                ++_Niter;
+            }
+        }
+
+        auto _Product = _Get_random_product(_Index, _Niter);
+        auto _Rem     = static_cast<_Udiff>(_Product) & _Mask;
+
+        if (_Rem < _Index) {
+            const auto _Threshold = (_Mask - (_Index - 1)) % _Index;
+            while (_Rem < _Threshold) {
+                _Product = _Get_random_product(_Index, _Niter);
+                _Rem     = static_cast<_Udiff>(_Product) & _Mask;
+            }
+        }
+
+        unsigned int _Generated_bits;
+        if constexpr (_Bits < _Udiff_bits) {
+            _Generated_bits = static_cast<unsigned int>(_Popcount(_Mask));
+        } else {
+            _Generated_bits = _Udiff_bits;
+        }
+
+        return static_cast<_Diff>(_Product >> _Generated_bits);
+    }
+
+    _Udiff _Get_all_bits() {
+        _Udiff _Ret = _Get_bits();
+
+        if constexpr (_Bits < _Udiff_bits) {
+            for (unsigned int _Num = _Bits; _Num < _Udiff_bits; _Num += _Bits) { // don't mask away any bits

```

Comment:
This does mask some bits (specifically the top bits of the first call to `_Get_bits()` if `_Bits` does not divide `_Udiff_bits` without residue; am I misunderstanding the comment?

Reply:
This is preexisting from the original `_Rng_from_urng` class, but I think this comment is meant to highlight the difference from `operator()`. This function generates enough bits to completely fill a `_Udiff` integer, though as you say any excess bits will be truncated. In contrast, `operator()` generates bits until the maximum possible random number is greater than or equal to `_Index-1`, which might be fewer bits than a `_Udiff` can store. In that sense, it's statistically equivalent to generating a full-width `_Udiff` random integer and then masking it, but without actually doing so.

Reply:
Yep, that was the intended meaning of the comment (which I originally wrote).